pkg/controller/propagator/propagation.go

```diff
@@ -350,3 +387,72 @@ func (r *ReconcilePolicy) handleDecision(instance *policiesv1.Policy, decision a
 	}
 	return nil
 }
+
+// a helper to quickly check if there are any templates in any of the policy templates
+func policyHasTemplates(instance *policiesv1.Policy) bool {
+	for _, policyT := range instance.Spec.PolicyTemplates {
+		if templates.HasTemplate(policyT.ObjectDefinition.Raw, templateCfg.StartDelim) {
+			return true
+		}
+	}
+	return false
+}
+
+func processTemplates(replicatedPlc *policiesv1.Policy, decision appsv1.PlacementDecision, rootPlcNamespace string) error {
+	reqLogger := log.WithValues("Policy-Namespace", replicatedPlc.GetNamespace(), "Policy-Name", replicatedPlc.GetName(),
+		"Managed-Cluster", decision.ClusterName, "Root Policy-Namespace", rootPlcNamespace)
+	reqLogger.Info("Processing Templates..")
+
+	templateCfg.LookupNamespace = rootPlcNamespace
+	tmplResolver, err := templates.NewResolver(kubeClient, kubeConfig, templateCfg)
+	if err != nil {
+		reqLogger.Error(err, "Error fetching template resolver")
+		return err
+	}
+
+	//A policy can have multiple policy templates within it, iterate and process each
+	for _, policyT := range replicatedPlc.Spec.PolicyTemplates {
```

Comment:
Since you previously mentioned you were concerned about performance. I'd suggest running the code inside of this for loop as individual go routines. The code in there appears to be thread-safe.